src/coreclr/src/vm/threadsuspend.cpp

```diff
@@ -3586,34 +3583,22 @@ HRESULT ThreadSuspend::SuspendRuntime(ThreadSuspend::SUSPEND_REASON reason)
             //     - Otherwise, we rely on the GCPOLL mechanism enabled by
             //       TrapReturningThreads.
 
-            // When reading shared state we need to erect appropriate memory barriers.
-            // The interlocked operation below ensures that any future reads on this
-            // thread will happen after any earlier writes on a different thread.
-            //
-            // <TODO> Need more careful review of this </TODO>
-            //
-            FastInterlockOr(&thread->m_fPreemptiveGCDisabled, 0);
-
-            if (thread->m_fPreemptiveGCDisabled)
-            {
-                FastInterlockOr((ULONG *) &thread->m_State, Thread::TS_GCSuspendPending);
-                countThreads++;
-
 #if defined(FEATURE_HIJACK) && defined(TARGET_UNIX)
-                bool gcSuspensionSignalSuccess = thread->InjectGcSuspension();
-                if (!gcSuspensionSignalSuccess)
-                {
-                    STRESS_LOG1(LF_SYNC, LL_INFO1000, "Thread::SuspendRuntime() -   Failed to raise GC suspension signal for thread %p.\n", thread);
-                }
-#endif // FEATURE_HIJACK && TARGET_UNIX
+            bool gcSuspensionSignalSuccess = thread->InjectGcSuspension();
```

Comment:
Review: if suspension signal inevitably leads the target thread to suspension, we could use  `TS_GCSuspendRedirected` also on Unix. Think about this.

Reply:
It does not look like `TS_GCSuspendRedirected` could be useful on Unix. 

On one hand - the signal handler does not guarantee that the thread will be suspended. So we can't consider that we are done with the thread after just posting a signal.
When the handler runs it will do a similar choice as here - whether thread is in managed/unmanaged/interruptible code and act accordingly. I.E. it may do a hijack or do nothing at all (in unmanaged case).

Also if the thread is in interruptible GC state, there is no redirection, handler just pushes redirection frame with the current context and pulses GC mode. I guess the reason why we do redirection on Windows (to get usable context) does not exist on Unix.

Reply:
>  I guess the reason why we do redirection on Windows (to get usable context) does not exist on Unix.

@janvorli  - to confirm.

Reply:
Right. That is kind of what I meant. 

I am referring to the comment:
```
                        // Redirect thread so we can capture a good thread context
                        // (GetThreadContext is not sufficient, due to an OS bug).
                        if (!thread->CheckForAndDoRedirectForGC())
```

On Windows, when thread is stopped at interruptible place, we cannot just get its context, we need to redirect it to a function that can get the context and then we release the thread so that it runs that function. At that point we just need to wait until the thread goes preemptive. Resuming the thread may take time though and if we see no other progress we may decide to retry suspending, but it would be a waste for already redirected threads since they are already in a state that will lead to suspension. Adding additional state `TS_GCSuspendRedirected` helps with that.

This is not a scenario on Unix. Since there is no redirection, there is no need to improve it.

Reply:
The question came up in code review - "Would it make sense to do the same on Unix?", and the answer is "No".

Reply:
The comment is misleading. The purpose of redirection is not to get a good context on Windows. The `CheckForAndDoRedirectForGC` itself checks if the context is good (using `Thread::IsContextSafeToRedirect`)and returns false if it is not good. Only if the context was good, we do the redirection